spec/client/ics-008-wasm-client/README.md

```diff
@@ -84,329 +111,457 @@ interface Height {
+
+  // initialize wasm contract for a previously stored contract identified by codeHash
+  initContract(codeHash, clientStore, marshalJSON(payload))
+
+  return ClientState{
+    data,
+    codeHash,
+    latestHeight: height
+  }
 }
 ```
 
-### Misbehaviour
+### Contract payload messages
 
-The `Misbehaviour` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
-Wasm client `Misbehaviour` consists of two conflicting headers both of which the light client would have considered valid.
+The Wasm client proxy performs calls to the Wasm client via the Wasm VM. The calls require as input payload messages that are categorized on two discriminated union types: one for payload messages used in calls that perform only reads, and one for payload messages used in calls that perform state-changing writes.
 
 ```typescript
-interface Misbehaviour {
-  clientId: string
-  h1: Header
-  h2: Header
-}
+type QueryMsg =
+  | Status 
+  | ExportMetadata
+  | TimestampAtHeight
+  | VerifyClientMessage
+  | VerifyMembership
+  | VerifyNonMembership
+  | CheckForMisbehaviour;
 ```
 
-### Client initialisation
+``` typescript
+type SudoMsg =
+  | UpdateState
+  | UpdateStateOnMisbehaviour
+  | VerifyUpgradeAndUpdateState
+  | CheckSubstituteAndUpdateState
+```
+
+### Validity predicate
+
+Wasm client validity checking uses underlying Wasm contract. If the provided client message is valid, the client state will proceed to checking for misbehaviour (call to `checkForMisbehaviour`) and updating state (call to either `updateStateOnMisbehaviour` or `updateState` depending on whether misbehaviour was detected in the client message). 
 
-Wasm client initialisation requires a (subjectively chosen) latest consensus state interpretable by the target Wasm Client Code. `wasmCodeId` field is unique identifier for `Wasm Client Code`, and `initializationData` refers to opaque data required for initialization of the particular client managed by that Wasm Client Code.
+```typescript
+interface VerifyClientMessageMsg {
+  clientMessage: ClientMessage
+}
+```
 
 ```typescript
-function initialise(
-    wasmCodeId: String,
-    initializationData: []byte,
-    consensusState: []byte,
-  ): ClientState {
-    codeHandle = getWasmCode(wasmCodeId)
-    assert(codeHandle.isInitializationDataValid(initializationData, consensusState))
+function verifyClientMessage(clientMsg: ClientMessage) {
+  payload = verifyClientMessageMsg{clientMsg}
 
-    store = getStore("clients/{identifier}")
-    return codeHandle.initialise(store, initializationData, consensusState)
+  clientState = provableStore.get("clients/{clientMsg.identifier}/clientState")
+  // retrieve client identifier-prefixed store
+  clientStore = provableStore.prefixStore("clients/{clientMsg.identifier}")
+
+  // use underlying wasm contract to verify client message
+  assert(callContract(clientStore, clientState, marshalJSON(payload)))
```

Comment:
NIT: Since the read-only operations are now changed to be queries, I would use `queryContract` when a read-only query is made to the underlying contract and `callContract` when an execute/sudo entrypoint is called.

Reply:
Thanks for the feedback. Yes, I agree; I will change that.